File: ppt_gen/mermaid.py
from __future__ import annotations

import json
import shutil
import subprocess
from pathlib import Path

from ppt_gen.paths import MERMAID_DIR, ROOT, theme_dir
from ppt_gen.theme.tokens import ThemeTokens


def mermaid_source_path(name: str) -> Path:
    path = MERMAID_DIR / f"{name}.mmd"
    if not path.exists():
        raise FileNotFoundError(f"Mermaid diagram not found: {path}")
    return path


def _mmdc_bin() -> str | None:
    local = ROOT / "node_modules" / ".bin" / "mmdc"
    if local.exists():
        return str(local)
    return shutil.which("mmdc")


def wrap_mermaid_source(source: str, tokens: ThemeTokens) -> str:
    mermaid_json = theme_dir(tokens.name) / "mermaid.json"
    if mermaid_json.exists():
        config = json.loads(mermaid_json.read_text())
    else:
        config = {"theme": "base", "themeVariables": {}}
    init = {
        "theme": config.get("theme", "base"),
        "themeVariables": config.get("themeVariables", {}),
    }
    init_line = f"%%{{init: {json.dumps(init)}}}%%\n"
    if source.strip().startswith("%%{init"):
        return source
    return init_line + source


def render_mermaid(
    name: str,
    tokens: ThemeTokens,
    output_path: Path,
) -> Path:
    source_path = mermaid_source_path(name)
    wrapped = wrap_mermaid_source(source_path.read_text(), tokens)
    output_path.parent.mkdir(parents=True, exist_ok=True)

    mmdc = _mmdc_bin()
    if not mmdc:
        raise RuntimeError(
            "mmdc not found. Run: npm install\n"
            "(@mermaid-js/mermaid-cli is listed in package.json devDependencies)"
        )

    tmp = output_path.with_suffix(".mmd.tmp")
    tmp.write_text(wrapped)
    try:
        result = subprocess.run(
            [
                mmdc,
                "-i",
                str(tmp),
                "-o",
                str(output_path),
                "-b",
                tokens.facecolor,
            ],
            check=False,
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            stderr = result.stderr.strip() or result.stdout.strip()
            raise RuntimeError(f"mmdc failed for diagram '{name}': {stderr}")
    finally:
        tmp.unlink(missing_ok=True)

    return output_path

File: lib/mermaid.mjs
import { existsSync } from "node:fs";
import { mkdir, readFile, unlink, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { spawnSync } from "node:child_process";

import { MERMAID_DIR, ROOT, themeDir } from "./paths.mjs";

export function resolveChromePath() {
  if (process.env.PUPPETEER_EXECUTABLE_PATH) {
    return process.env.PUPPETEER_EXECUTABLE_PATH;
  }
  const candidates = [
    "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
    "/Applications/Chromium.app/Contents/MacOS/Chromium",
    "/usr/bin/google-chrome",
    "/usr/bin/chromium",
    "/usr/bin/chromium-browser",
  ];
  return candidates.find((p) => existsSync(p)) ?? null;
}

export function browserEnv() {
  const chrome = resolveChromePath();
  const env = { ...process.env };
  if (chrome) {
    env.PUPPETEER_EXECUTABLE_PATH = chrome;
    env.CHROME_PATH = chrome;
  }
  return env;
}

export function mermaidSourcePath(name) {
  const path = join(MERMAID_DIR, `${name}.mmd`);
  if (!existsSync(path)) {
    throw new Error(`Mermaid diagram not found: ${path}`);
  }
  return path;
}

function mmdcBin() {
  const local = join(ROOT, "node_modules", ".bin", "mmdc");
  if (existsSync(local)) return local;
  return null;
}

export async function wrapMermaidSource(source, tokens) {
  const mermaidJson = join(themeDir(tokens.name), "mermaid.json");
  let config = { theme: "base", themeVariables: {} };
  if (existsSync(mermaidJson)) {
    config = JSON.parse(await readFile(mermaidJson, "utf8"));
  }
  const init = {
    theme: config.theme ?? "base",
    themeVariables: config.themeVariables ?? {},
  };
  if (source.trim().startsWith("%%{init")) return source;
  return `%%{init: ${JSON.stringify(init)}}%%\n${source}`;
}

export async function renderMermaid(name, tokens, outputPath) {
  const sourcePath = mermaidSourcePath(name);
  const wrapped = await wrapMermaidSource(
    await readFile(sourcePath, "utf8"),
    tokens,
  );
  await mkdir(dirname(outputPath), { recursive: true });

  const mmdc = mmdcBin();
  if (!mmdc) {
    throw new Error(
      "mmdc not found. Run: npm install\n(@mermaid-js/mermaid-cli is listed in package.json)",
    );
  }

  const tmp = `${outputPath}.mmd.tmp`;
  await writeFile(tmp, wrapped);
  try {
    const result = spawnSync(
      mmdc,
      ["-i", tmp, "-o", outputPath, "-b", tokens.facecolor],
      { encoding: "utf8", cwd: ROOT, env: browserEnv() },
    );
    if (result.status !== 0) {
      const stderr = (result.stderr || result.stdout || "").trim();
      throw new Error(`mmdc failed for diagram '${name}': ${stderr}`);
    }
  } finally {
    await unlink(tmp).catch(() => {});
  }

  return outputPath;
}
